File: java/examples/DTRGenerator.java
//
// The nordugrid-arc-java package is required. To compile and run this example:
// 
// export CLASSPATH=/usr/lib64/java/arc.jar:.
// export LD_LIBRARY_PATH=/usr/lib64/java
// javac DTRGenerator.java
// java DTRGenerator /bin/ls /tmp/dtrtest
// 
// The PATHs above may vary depending on ARC install location and system
// architecture. 

import nordugrid.arc.*; // For the sake of brevity in this example import everything from arc
import com.sun.security.auth.module.UnixSystem; // to get uid

// Implementation of DTR Generator.
// Cannot inherit from DTRCallback as it is a pure virtual class and swig does not
// create a default constructor, so extend Scheduler which inherits from DTRCallback.
class DTRGenerator extends Scheduler {
    private Logger logger;
    private LogDestination logdest;
    private SimpleCondition cond;

    // Create a new Generator and set up logging to stdout
    public DTRGenerator() {
        logger = new Logger(Logger.getRootLogger(), "Generator");
        logdest = new LogStream_ostream(arc.getStdout());
        Logger.getRootLogger().addDestination(logdest);
        Logger.getRootLogger().setThreshold(LogLevel.DEBUG);
        cond = new SimpleCondition();
    }

    // Implementation of callback from DTRCallback
    public void receiveDTR(DTRPointer dtr) {
        // root logger is disabled in Scheduler thread so need to add it here
        Logger.getRootLogger().addDestination(logdest);
        logger.msg(LogLevel.INFO, "Received DTR " + dtr.get_id() + " in state " + dtr.get_status().str());
        Logger.getRootLogger().removeDestinations();
        cond.signal();
    }

    // Run the transfer and wait for the callback on completion
    private void run(final String source, final String dest) {
        // Set log level for DTR (must be done before starting Scheduler)
        DTR.setLOG_LEVEL(LogLevel.DEBUG);

        // Start Scheduler thread
        Scheduler scheduler = new Scheduler();
        scheduler.start();

        // UserConfig contains information such as the location of credentials
        UserConfig cfg = new UserConfig();

        // The ID can be used to group DTRs together
        String id = "1234";

        // Logger for DTRs
        DTRLogger dtrlog = arc.createDTRLogger(Logger.getRootLogger(), "DTR");
        dtrlog.addDestination(logdest);

        // Create a DTR
        UnixSystem unix = new UnixSystem();
        DTRPointer dtr = arc.createDTRPtr(source, dest, cfg, id, (int)unix.getUid(), dtrlog);
        logger.msg(LogLevel.INFO, "Created DTR "+ dtr.get_id());

        // Register this callback in order to receive completed DTRs
        dtr.registerCallback(this, StagingProcesses.GENERATOR);
        // This line must be here in order to pass the DTR to the Scheduler
        dtr.registerCallback(scheduler, StagingProcesses.SCHEDULER);

        // Push the DTR to the Scheduler
        DTR.push(dtr, StagingProcesses.SCHEDULER);

        // Wait until callback is called
        // Note: SimpleCondition.wait() is renamed to _wait() as wait() is a java.lang.Object method
        cond._wait();

        // DTR is finished, so stop Scheduler
        scheduler.stop();
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.out.println("Usage: java DTRGenerator source destination");
            return;
        }
        DTRGenerator gen = new DTRGenerator();
        gen.run(args[0], args[1]);
    }
}

File: python/examples/dtr_generator.py
#!/usr/bin/python2

# The nordugrid-arc-python package is required. As stated in the comments, the
# main missing piece in Python when compared to C++ is the ability to get
# callbacks to the Generator when the DTR has finished. To run:
#
# python dtr_generator.py /bin/ls /tmp/dtrtest
#
# If nordugrid-arc-python is installed to a non-standard location, PYTHONPATH
# may need to be set.

import os
import sys
import time
import arc

class DTRGenerator(arc.DTRCallback):

    def __init__(self):
        super(DTRGenerator, self).__init__()
        # Set up logging
        self.root_logger = arc.Logger_getRootLogger()
        self.stream = arc.LogStream(sys.stdout)
        self.root_logger.addDestination(self.stream)
        self.root_logger.setThreshold(arc.DEBUG)
        self.cfg = arc.UserConfig('', '')
        self.id = '1'
        arc.DTR.LOG_LEVEL = self.root_logger.getThreshold()

        # Start the Scheduler
        self.scheduler = arc.Scheduler()
        self.scheduler.start()

    def __del__(self):
        # Stop Scheduler when Generator is finished
        self.scheduler.stop()

    def add(self, source, dest):
        # Logger object, wrapped in smart pointer. The Logger object can only be accessed
        # by explicitly deferencing the smart pointer.
        dtrlog = arc.createDTRLogger(self.root_logger, "DTR")
        dtrlog.__deref__().addDestination(self.stream)
        dtrlog.__deref__().setThreshold(arc.DEBUG)

        # Create DTR (also wrapped in smart pointer)
        dtrptr = arc.createDTRPtr(source, dest, self.cfg, self.id, os.getuid(), dtrlog)

        # The ability to register 'this' as a callback object is not available yet
        #dtrptr.registerCallback(self, arc.GENERATOR)
        # Register the scheduler callback so we can push the DTR to it
        dtrptr.registerCallback(self.scheduler, arc.SCHEDULER)
        # Send the DTR to the Scheduler
        arc.DTR.push(dtrptr, arc.SCHEDULER)
        # Since the callback is not available, wait until the transfer reaches a final state
        while dtrptr.get_status() != arc.DTRStatus.ERROR and dtrptr.get_status() != arc.DTRStatus.DONE:
            time.sleep(1)
        sys.stdout.write("%s\n"%dtrptr.get_status().str())

    # This is never called in the current version
    def receiveDTR(self, dtr):
        sys.stdout.write('Received back DTR %s\n'%str(dtr.get_id()))

def main(args):
    if len(args) != 3:
        sys.stdout.write("Usage: python dtr_generator.py source destination\n")
        sys.exit(1)
    generator = DTRGenerator()
    generator.add(args[1], args[2])

if __name__ == '__main__':
    main(sys.argv[0:])
